File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.cpp
// we can check every index on s to see if it includes p
// O(m*n)
// if s is of size m and if p is of size of n

// if p.len > s.len, we know it's impossible

// we can do a sliding window method
// we count the freq in p
// and traverse s using sliding window
// O(max(m*O(1), n))


class Solution {
public:


    vector<int> findAnagrams(string s, string p) {

        int m = s.length();
        int n = p.length();

        if (n > m)
            return {};

        // lowercase English letters.
        vector<int> cnts(26);
        for (const auto & c : p) 
            cnts[c - 'a']++;

        vector<int> ans;
        vector<int> currCnts(26);
        int l = 0;
        for (int r = 0; r < m; r++) {
            currCnts[s[r] - 'a']++;

            // don't need to remove by charater, just simply shrink to the same size
            // while (currCnts[s[r] - 'a'] > cnts[s[r] - 'a']) {
            //     currCnts[s[l] - 'a']--;
            //     l++;
            // }

            // remove one letter 
            // from the left side of the window
            if (r - l + 1 > n) {
                currCnts[s[l] - 'a']--;
                l++;
            }

            bool isAnagram = true;
            for (int i = 0; i < 26; i++) {
                if (currCnts[i] != cnts[i])
                    isAnagram = false;
            }
            if (isAnagram)
                ans.push_back(l);

        }
        

        return ans;
        
    }

};

File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        if len(p) > len(s):
            return []        
        
        sCnts = [0] * 26
        pCnts = [0] * 26
        ans = []
        
        for c in p:
            pCnts[ord(c) - ord('a')] += 1
        
        left = 0;
        for right in range(len(s)):
            # contract
            if right - left + 1 > len(p):
                sCnts[ord(s[left]) - ord('a')] -= 1
                left += 1
            
            sCnts[ord(s[right]) - ord('a')] += 1
            if sCnts == pCnts:
                ans.append(left)
        return ans
        